File: tools/sdks/go/examples/quickstart_byok_rotate.go
//go:build ignore

// quickstart_byok_rotate — File a BYOK rotation op via POST /v1/admin/ops.
//
// Customer concept: privileged ops are dual-approval gated
// (CTRL-DUAL-APPROVAL-001). PENDING until a second admin approves via
// POST /v1/admin/ops/{op_id}/approve.
//
// Run: CORELINK_PAT=$ADMIN_PAT go run examples/quickstart_byok_rotate.go
package main

import (
	"bytes"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"time"
)

func randomID() string {
	var b [16]byte
	if _, err := rand.Read(b[:]); err != nil {
		return "fallback-id"
	}
	return hex.EncodeToString(b[:])
}

func main() {
	api := os.Getenv("CORELINK_API_URL")
	if api == "" {
		api = "https://corelink-api.humangr.com"
	}
	pat := os.Getenv("CORELINK_PAT")
	if pat == "" {
		fmt.Fprintln(os.Stderr, "error: CORELINK_PAT env var is required")
		os.Exit(2)
	}
	idem := "idem-" + randomID()
	body, err := json.Marshal(map[string]any{
		"op_type": "byok_rotate",
		"reason":  "scheduled quarterly rotation",
		"params": map[string]any{
			"key_alias":     "primary",
			"target_region": "eu-west-1",
		},
	})
	if err != nil {
		fmt.Fprintf(os.Stderr, "marshal: %v\n", err)
		os.Exit(1)
	}
	req, err := http.NewRequest("POST", api+"/v1/admin/ops", bytes.NewReader(body))
	if err != nil {
		fmt.Fprintf(os.Stderr, "build request: %v\n", err)
		os.Exit(1)
	}
	req.Header.Set("Authorization", "Bearer "+pat)
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Idempotency-Key", idem)
	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		fmt.Fprintf(os.Stderr, "http: %v\n", err)
		os.Exit(1)
	}
	defer resp.Body.Close()
	out, _ := io.ReadAll(resp.Body)
	fmt.Printf("status=%d body=%s\nNOTE: op is PENDING until a second admin approves it.\n", resp.StatusCode, string(out))
}

File: tools/cli/examples/quickstart_byok_rotate.rs
//! `quickstart_byok_rotate` — File a BYOK rotation op via `POST /v1/admin/ops`.
//!
//! Customer concept: privileged operations in CoreLink are
//! **dual-approval gated** (CTRL-DUAL-APPROVAL-001). Filing the op
//! returns an `op_id`; a second admin must approve via
//! `POST /v1/admin/ops/{op_id}/approve` before execution.
//!
//! Run:
//!
//! ```text
//! CORELINK_PAT=$ADMIN_PAT \
//!   cargo run --example quickstart_byok_rotate -p corelink-cli
//! ```

#![forbid(unsafe_code)]
#![allow(
    clippy::print_stdout,
    reason = "examples emit human-readable narration to stdout"
)]

use anyhow::{Context, Result};
use serde_json::json;
use std::env;
use uuid::Uuid;

#[tokio::main]
async fn main() -> Result<()> {
    let api = env::var("CORELINK_API_URL")
        .unwrap_or_else(|_| "https://corelink-api.humangr.com".to_string());
    let pat = env::var("CORELINK_PAT").context("CORELINK_PAT is required")?;
    let idem = format!("idem-{}", Uuid::now_v7());

    let body = json!({
        "op_type": "byok_rotate",
        "reason": "scheduled quarterly rotation",
        "params": { "key_alias": "primary", "target_region": "eu-west-1" },
    });

    let resp = reqwest::Client::new()
        .post(format!("{api}/v1/admin/ops"))
        .bearer_auth(&pat)
        .header("Idempotency-Key", &idem)
        .json(&body)
        .send()
        .await
        .context("POST /v1/admin/ops")?;

    let status = resp.status();
    let text = resp.text().await.context("read response body")?;
    println!("status={status} body={text}");
    println!("NOTE: op is PENDING until a second admin approves it.");
    Ok(())
}
